theano/scan_module/tests/test_scan.py

```diff
@@ -2519,6 +2519,53 @@ def f_rnn(u_t, x1_tm1, x1_tm3, x2_tm1, x3tm2, x3_tm1, x4_tm1):
         utt.assert_allclose(tx4, v_u[-1] + 4.)
         utt.assert_allclose(tx5, v_u[-1] + 5.)
 
+    def test_use_scan_direct_output(self):
+        # This test looks for a crash that happened when directly using the
+        # recurrent output of a scan node instead of taking the result
+        # returned by the scan() function
+
+        # Obtain a compilation mode that will cause the test to fail if an
+        # exception occurs in the optimization process
+        on_opt_error = theano.config.on_opt_error
+        theano.config.on_opt_error = "raise"
+        mode = theano.compile.get_default_mode()
+        theano.config.on_opt_error = on_opt_error
```

Comment:
Use `try / finally` so that the value is restored even if an exception is raised in `get_default_mode()` (however unlikely that is).